### PR TITLE
feat: Allow setting the `SLURMEnvironment.main_address` via an env variable

### LINKED ISSUE #17550: Concurrent distributed training runs on a SLURM cluster with `srun`
### Description & Motivation

When doing a parameter sweep (HPO), you often want to run multiple training runs concurrently. In our use case, the training runs use the `ddp` strategy (1 node, 4 GPUs) and are executed on a SLURM cluster.

We reserve the required resources via `sbatch` and then run the single training runs via `srun`. With this approach, we cannot use the [SLURM ClusterEnvironment](https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/environments/slurm.py#L31) since the [main address](https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/environments/slurm.py#L56) is determined by the first entry of the `SLURM_NODELIST`, which is the list of nodes for the whole job (sbatch), not the job step (srun).

From a single training run, we saw that using the SLURM ClusterEnvironment is a bit faster, which is also expected. So we would like to use it for our parameter sweeps as well.

### Pitch

We would like to be able to specify the main address manually when creating a [SLURMEnvironment](https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/environments/slurm.py#L31) instance, something like:
```python
concurrent_slurm_env = SLURMEnvironment(main_address="localhost")
trainer = Trainer(plugins=concurrent_slurm_env)
```

### Alternatives

Alternatively one could try to extract the main address from the `SLURM_STEP_NODELIST` env variable (not available on all SLURM clusters), or via the `scontrol` command.

### Additional context

If you think this feature request is reasonable, I would be more than happy to provide a PR.

cc @borda @awaelchli

### PR DESCRIPTION
**Pull request by dcfidalgo**
## What does this PR do?

Fixes #17550 
With this PR the user is able to configure the master address of the process group when running a DDP training on a SLURM cluster. This makes it possible to run concurrent training runs within a SLURM job, enabling efficient parameter sweeps/tunings.

In #17550 we discussed adding an argument to the constructor of the `SLURMEnvironment` class, but I think it is cleaner (more consistent) to use a environment variable for that:
- the main port is already configurable through an [env variable](https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/environments/slurm.py#L81)
- it is the canonical way in PyTorch Distributed (see [here](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#basic-use-case), for example)

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]
Looks great, thanks @dcfidalgo! FYI, I moved the test to the right file and used monkeypatch to set the environment variable.

**Review by carmocca — APPROVED** [review]
